components/playlists/browser/playlists_media_file_controller.cc

```diff
@@ -0,0 +1,342 @@
+  url_loaders_.clear();
+}
+
+void PlaylistsMediaFileController::NotifyFail() {
+  ResetDownloadStatus();
+  client_->OnMediaFileGenerationFailed(std::move(current_playlist_));
+}
+
+void PlaylistsMediaFileController::NotifySucceed() {
+  ResetDownloadStatus();
+  client_->OnMediaFileReady(std::move(current_playlist_));
+}
+
+void PlaylistsMediaFileController::DeletePlaylist(
+    const base::FilePath& path) {
+  io_task_runner()->PostTask(FROM_HERE,
+                             base::BindOnce(&DeleteFile, path));
+}
+
+void PlaylistsMediaFileController::GenerateSingleMediaFile(
+    base::Value&& playlist_value,
+    const base::FilePath& base_dir) {
+  DCHECK(!in_progress_);
+
+  in_progress_ = true;
+  current_playlist_ = std::move(playlist_value);
+  current_playlist_id_ = *current_playlist_.FindStringKey("id");
+  remained_download_files_ = GetNumberOfMediaFileSources();
+  media_file_source_files_count_ = remained_download_files_;
+  playlist_dir_path_ =
+      base_dir.Append(GetPlaylistIDDirName(current_playlist_id_));
+
+  // Create PROFILE_DIR/playlists/ID/source_files dir to store each media files.
+  // Then, downloads them in that directory.
+  CreateSourceFilesDirThenDownloads();
+}
+
+void PlaylistsMediaFileController::CreateSourceFilesDirThenDownloads() {
+  const base::FilePath source_files_dir =
+      playlist_dir_path_.Append(kSourceMediaFilesDir);
+  base::PostTaskAndReplyWithResult(
+      io_task_runner(),
+      FROM_HERE,
+      base::BindOnce(&EnsureDirectoryExists, source_files_dir),
+      base::BindOnce(&PlaylistsMediaFileController::OnSourceFilesDirCreated,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void PlaylistsMediaFileController::OnSourceFilesDirCreated(bool success) {
+  if (!success) {
+    NotifyFail();
+    return;
+  }
+
+  DownloadAllMediaFileSources();
+}
+
+int PlaylistsMediaFileController::GetNumberOfMediaFileSources() {
+  DCHECK(in_progress_);
+  base::Value* media_files =
+      current_playlist_.FindPath("createParams.mediaFiles");
+  return media_files->GetList().size();
+}
+
+void PlaylistsMediaFileController::DownloadAllMediaFileSources() {
+  base::Value* media_files =
+      current_playlist_.FindPath("createParams.mediaFiles");
+  for (int i = 0; i < media_file_source_files_count_; ++i) {
+    const std::string* url =
+        media_files->GetList()[i].FindStringKey("mediaFileUrl");
+    if (url) {
+      DownloadMediaFile(GURL(*url), i);
+    } else {
+       NOTREACHED() << "Playlists has Empty media file url";
+       NotifyFail();
+       break;
+    }
+  }
+}
+
+void PlaylistsMediaFileController::DownloadMediaFile(const GURL& url,
+                                                     int index) {
+  LOG(ERROR) << __func__ << ": " << url.spec() << " at: " << index;
+
+  auto request = std::make_unique<network::ResourceRequest>();
+  request->url = GURL(url);
+  request->allow_credentials = false;
+  network::SimpleURLLoader* loader = network::SimpleURLLoader::Create(
```

Comment:
see comments in `DownloadThumbnail`, this is not a raw pointer, and fixing it leads to an annoying cascade of other changes

Reply:
Fixed.